# 1. Introduction to Python

Python is a programming language.
It is not a mathematics-oriented language in and of itself.
It is a general-purpose language, meaning we can do pretty much what we want with it.

[![Python](https://imgs.xkcd.com/comics/python.png)](https://xkcd.com/353/)

Here is a (supershort) list of what humanity did with Python:

- Dropbox (Source: [Dropbox Blog](https://blogs.dropbox.com/tech/2018/09/how-we-rolled-out-one-of-the-largest-python-3-migrations-ever/))
- Image editing ([The GNU Image Manipulation Program](https://www.gimp.org/))
- Vector graphics ([Inkscape](https://inkscape.org/))
- 3D modeling ([Blender](https://www.blender.org/))
- Desktop publishing ([Scribus](https://www.scribus.net/))
- Web pages ([Reddit](https://www.reddit.com/), Source: [Reddit Blog](https://redditblog.com/2005/12/05/on-lisp/))

Also in economics we have a great sites like [QuantEcon](https://quantecon.org) by **Prof. Sargent** and his team and you can find many economics models on it.

# 2. How to work with the notebooks in colab/binder

## Jupyter notebook:

- notebook as a collection of cells
- two main cell types: text (markdown) and code
- click on a cell to edit the contents
- Shift+Enter to run the code within a code cell or render the text cell
- be aware of the execution order
- installing necessary modules with pip and conda



This is the pyhton course

In [ ]:
# hotkey

In [ ]:
# markdown

In [6]:
z = 2
z * 3

6

$$ \alpha = \beta = \frac{\gamma^3}{\sqrt{\theta}}$$

Every code line in Python is interpreted as a `command`, unless it starts
with the hash/pound sign, in which case it is considered to be a `comment`:

```
# the line below will be executed
1+1

# the line below will not be executed
# 1+2
```

In [7]:
r = 0.03 # interest rate

In [9]:
# a = 2

# 3. The Basics

Python alone cannot do much.
For this reason, we are almost always going to work with a package (It will be taught in future sessions).
However, it is fundamental to understand the basics.
This involves familiarizing with the _syntax_ and with the basic _data types_.

Syntax is the set of rules that govern writing code.
This includes how to write an _assignment_ (providing a variable with a value), how to call functions and how to access items in a iterable object (e.g., lists, arrays).
It also includes _code blocks,_ which execute conditionally on a given rule (e.g., `if`, `while`).



## 3.1 Arithmetic operations and variables

Let's start with basic calculations in Python. Entering some arithmetic operation in the code cell below (e.g. `2+2`):

<br><center> <b>Arithmetic Operators<b> <center>
<br>

| Symbol | Task Performed | Description| Example<br> (a = 10 & b = 20) |
|:----:|:---:|:---:|:---:| 
| +  | Addition | Adds values on either side of the operator. | a + b = 30
| -  | Subtraction | Subtracts right hand operand from left hand operand. | a – b = -10
| /  | division | Divides left hand operand by right hand operand | b / a = 2
| %  | mod | Divides left hand operand by right hand operand and returns remainder | b % a = 0
| *  | multiplication | Multiplies values on either side of the operator | a * b = 200
| //  | floor division | he division of operands where the result is the quotient in<br> which the digits after the decimal point are removed. | b // a = 2
| **  | to the power of | Performs exponential (power) calculation on operators | a**b =$10^{20}$

In [10]:
2 + 4

6

In [11]:
6 - 4

2

In [12]:
6 / 4

1.5

In [13]:
6 * 12

72

In [14]:
6 % 4

2

In [15]:
6 // 4

1

In [20]:
2 ** 10

1024

In [22]:
5 + (4 - 3*2)**3 + 1

-2

As with many programming languages, we are defining variables and changing their values all the time in Python.
We can create a variable simply by inventing a name and assigning a value to it, like the following.

To store the results of a calculation in a `variable` we can use `=` sign.

```python
sample_number = 2 + 3
```

When working inside the notebooks, we can see the result of the last executed command
or a `variable`. For example, what do you see when you type the code below in a blank cell?

```
sample_number
```

In [23]:
a = 5 + (4 - 3*2)**3 + 1

In [24]:
a

-2

Later in the code we can re-use this variable, check what is the output of this command:

```python
sample_number + 1
```

In [25]:
a + 3

1

In [29]:
a = 5 + (4 - 3*2)**3 + 1
a = 100
a = 20
a = a + 10

a

30

## 3.2 Relational Operators

Relational operators are used for comparing the values. It either returns **True** or **False** according to the condition. These operators are also known as Comparison Operators.

| Symbol | Task Performed | Description|
|:----:  |:---:|:---:| 
|=	     |Assignment| Assigns values from right side operands to left side operand
|==      |True, if it is equal| If the values of two operands are equal, then the condition becomes true.|
|!=      |True, if not equal to| If values of two operands are not equal, then condition becomes true.|
|<	     |less than| If the value of left operand is less than the value of right operand,<br> then condition becomes true.
|>	     |greater than| If the value of left operand is greater than the value of right operand, <br>then condition becomes true.
|<=      |less than or equal to| If the value of left operand is less than or equal to the value of right operand,<br> then condition becomes true.
|>=      |greater than or equal to|If the value of left operand is greater than or equal to the value of right operand, <br>then condition becomes true.

In [30]:
x = 2
y = 6

In [34]:
print(x==y)
print(2==6)
print(x==2)
print(x!=y)

False
False
True
True


In [37]:
print(x>y)
print(x>2)
print(x>=2)

False
False
True


In [38]:
x > 1 and y < 10

True

In [39]:
x > 1 and y > 10

False

In [40]:
x > 1 or y > 10

True

## 3.3 Type

Now we have a variable whose name is `a` and its value is `1`.
This variable has a _type_, which is the kind of value it contains.
In particular, `1` is an integer, whose type is represented in Python with the keyword `int`.
Understanding that variables can be of different types is important, because the type defines what we can do with that variable.
We can ask Python to tell us the type of a variable by using the `type()` function.

Data types are the types of variables we can create.
Python defines a few basic ones and packages (see below) provide new data types.
This programming language features _dynamic typing_, which means that we do not have to define what a variable can be.
Instead, Python infers the type of a variable when we create it.
Examples of basic data types are strings (`str`), lists (`list`), integer numbers (`int`).
<br><br>

| Types | Example |
| :--: | :--: |
| string | "Hello" ,  'World' |
| integer | 1 , 2 , 3 |
| float | 1.2 , 4.6 , 112.6 |
| boolian | True , Flase |
|list | [1, 2, 3, 'python', 9, 7] |
| dictionary | {'Python': 18 , "Econ": 20}



In [45]:
print(type('Economics'))
print(type(1))
print(type(1.3))
print(type(False))

<class 'str'>
<class 'int'>
<class 'float'>
<class 'bool'>


### 3.3.1 Text (`str`)

A basic object in Python is text.
This is formally represented with a `str` object.
Strings are denoted either with single or double quotes, and the choice between them is a matter of taste.

In [46]:
print("Hello world!")

Hello world!


There must always be consistency between opening a closing quotes. This means that single quotes can only close single quotes and double quotes can only close double quotes.

In [48]:
# "Weird string'

In [49]:
"python" + " " + 'course!'

'python course!'

In [50]:
print("python" + " " + 'course!')

python course!


In [51]:
a = 'python course'
a

'python course'

In [52]:
a.capitalize()

'Python course'

In [53]:
a

'python course'

In [54]:
a.upper()
# lower

'PYTHON COURSE'

In [55]:
b = 'Hi, this is python course!'
b

'Hi, this is python course!'

In [56]:
b.count('i')

3

In [59]:
b.find('t')

4

In [62]:
b + '2'

'Hi, this is python course!2'

In [63]:
b + str(2)

'Hi, this is python course!2'

### 3.3.2 Numbers (`int` and a`float`)

We already encountered integer numbers, whose type is `int`.
Another numerical type we are going to use very often is `float`.
This essentially is a non-integer real number, although the inner workings of [floating-point numbers](https://en.wikipedia.org/wiki/Floating-point_arithmetic) are more complicated.
We can initialize a variable to be of type `float` by simply assigning a decimal number to a variable.

In [64]:
type(123)

int

In [65]:
type(123.345)

float

This works even when the digit on the right of the decimal point is non significant.
The simple fact that we typed a period in the number tells Python that we want to work with floating point numbers.

In [66]:
type(1.0)

float

Both `int` and `float` variables support conventional arithmetic operations, such as addition (`+`), subtraction (`-`), multiplication (`*`), division (`/`) and raise to power (`**`).

Other operations that might be handy are floor division (`//`) and the _mod_ operation (`%`).
The former returns the largest integer smaller than the quotient, while the latter returns the remainder of floor division.

### 3.3.3. Iterable objects (`list`, `dict` and `tuple`)

Often we want to collect objects in arrays.
The most basic example is a vector, which is an organized array of numbers.
Python (mainly) provides three iterable objects.

####  3.3.3.1 **List**


We first look at **lists**.
These are arrays of heterogeneous objects.
They are created by collecting objects in square brackets.

A list of objects is an odered collection of items, for example `['a', 'b', 'c', 'a', 123]`. Since
lists are ordered, every item in the list has an associated index that starts with 0. This provides
access to specific elements of the list by specifying their index. For example, 'b' in the list above has index `1`.
The syntax for accessing a specific element is to provide its index in square brackets, e.g. `sample_list[1]`.

There are no restriction on the contents of the list, they can contain repeated values
(e.g. 'a' in the list above) or they can contain different data types (e.g. strings and integers in the list above).



In [68]:
print(type([]))
print(type([1]))

<class 'list'>
<class 'list'>


In [69]:
n = ['python', 4, 1.2]
n

['python', 4, 1.2]

In [70]:
type(n)

list

Lists can also nest.

In [72]:
['R', 234, n]

['R', 234, ['python', 4, 1.2]]

We can access the contents of any iterable object using square brackets.

In [73]:
x = [1,2,3,4,5,6,7,8,9,10]
x

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [74]:
x[0]

1

In [75]:
x[3]

4

In [76]:
x[-3]

8

Sometimes we want to use sequences inside a list, e.g. all items from start until the third item.
For this we can use the slicing notation:
```
[start:stop:step_size]
```

If `start` is missing, the assumption is to start from 0, if `stop` is missing the assumption
is to run up to and including the last item. If `step_size` is missing then iteration
will go over all items between `start` and `stop`. 

Specifying `step_size` will display every `step_size`'th item, e.g. `[::2]` will display every second
item starting from the first item. Note that specifying a negative number will reverse the order in which items in the list are
displayed. 

In [77]:
x

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [78]:
x[3:6]

[4, 5, 6]

In [79]:
x[2:-3]

[3, 4, 5, 6, 7]

In [81]:
x[:-3]

[1, 2, 3, 4, 5, 6, 7]

In [82]:
x[1:7]

[2, 3, 4, 5, 6, 7]

In [84]:
# [start:end:step]
x[1:7:3]

[2, 5]

In [86]:
x[::-1]

[10, 9, 8, 7, 6, 5, 4, 3, 2, 1]

In [87]:
x * 2

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [88]:
x

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [90]:
x + [4]

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 4]

- With `.append()`, you can add items to the end of an existing list object. However, you need to keep in mind that .append() adds only a single item or object at a time:

In [91]:
x

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [92]:
x.append(5)

In [93]:
x

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 5]

In [94]:
x.append([6,7])

In [95]:
x

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 5, [6, 7]]

- The `.extend()` method adds all the elements of an iterable (list, tuple, string etc.) to the end of the list.

In [96]:
x.extend([123, 541, 'Saeed'])

In [97]:
x

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 5, [6, 7], 123, 541, 'Saeed']

- The `.index()` method returns the index of the specified element in the list.
```python
list.insert(i, elm)
```
Here, `elem` is inserted to the list at the $i^{th}$ index. All the elements after `elem` are shifted to the right.

In [98]:
x

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 5, [6, 7], 123, 541, 'Saeed']

In [99]:
x.index('Saeed')

14

In [100]:
x.index(1)

0

In [101]:
x[14]

'Saeed'

In [102]:
x[0]

1

In [103]:
x.insert(3, 'Course')

In [104]:
x

[1, 2, 3, 'Course', 4, 5, 6, 7, 8, 9, 10, 5, [6, 7], 123, 541, 'Saeed']

In [105]:
x

[1, 2, 3, 'Course', 4, 5, 6, 7, 8, 9, 10, 5, [6, 7], 123, 541, 'Saeed']

- We have two methods (`.remove()` and `.pop()`) to delete elements in out lists.<br>
The `.remove(item)` method removes the first matching element (which is passed as an argument) from the list.  
The `.pop(index)` method removes the item at the given index from the list and returns the removed item.

In [106]:
x

[1, 2, 3, 'Course', 4, 5, 6, 7, 8, 9, 10, 5, [6, 7], 123, 541, 'Saeed']

In [107]:
x.remove('Course')  # item / value
x

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 5, [6, 7], 123, 541, 'Saeed']

In [108]:
x.pop(0)
x

[2, 3, 4, 5, 6, 7, 8, 9, 10, 5, [6, 7], 123, 541, 'Saeed']

In [109]:
x.pop(-1)   # index
x

[2, 3, 4, 5, 6, 7, 8, 9, 10, 5, [6, 7], 123, 541]

In [110]:
x

[2, 3, 4, 5, 6, 7, 8, 9, 10, 5, [6, 7], 123, 541]

In [111]:
x.reverse()
x

[541, 123, [6, 7], 5, 10, 9, 8, 7, 6, 5, 4, 3, 2]

In [112]:
x[::-1]

[2, 3, 4, 5, 6, 7, 8, 9, 10, 5, [6, 7], 123, 541]

- The `.sort()` method sorts the items of a list in ascending or descending order.<br>
The `.sort()` method accepts a reverse parameter as an optional argument.  
    - Setting `reverse = True` sorts the list in the descending order.


In [113]:
x = [1,2,3,4,10,3,54,-34]
x

[1, 2, 3, 4, 10, 3, 54, -34]

In [114]:
x.sort()
x

[-34, 1, 2, 3, 3, 4, 10, 54]

In [115]:
x.sort(reverse=True)
x

[54, 10, 4, 3, 3, 2, 1, -34]

####  3.3.3.2 **Dictionary**

Finally, we have **dictionaries**.
These are essentially lists, with the difference that each element is assigned to a _key_.
We create dictionaries using curly braces.


`Dictionary` is a data type that contains a combination of key-value items, for example:

```python
sample_dictionary = dict(math=19, econ=20, physics=16, geology=14)
sample_dictionary_alt = {'math' =19, 'econ'= 20, 'physics'= 16, 'geology'= 14}
print(sample_dictionary, sample_dictionary_alt)
```

Once a dictionary is defined we can access specific values by using relevant key, specifying
it in square brackets, for example:

```python
print(sample_dictionary['econ'])
```

We can use a similar notation to add new key/value pairs to the dictionary:

```python
sample_dictionary['python'] = True
print(sample_dictionary)
```

In [116]:
type({})

dict

In [ ]:
# dictionary

# { : , : , : } 

In [117]:
score = {'math':19, 'econ':20, 'geology':14, 'finance':18}
score

{'math': 19, 'econ': 20, 'geology': 14, 'finance': 18}

In [118]:
type(score)

dict

In [119]:
score['math']

19

In [120]:
score['econ']

20

In [121]:
score.keys()

dict_keys(['math', 'econ', 'geology', 'finance'])

In [122]:
score.values()

dict_values([19, 20, 14, 18])

In [123]:
score.items()

dict_items([('math', 19), ('econ', 20), ('geology', 14), ('finance', 18)])

In [124]:
score.get('math')

19

### tuple

In [125]:
a = (1,2,3)

In [126]:
type(a)

tuple

In [127]:
a[0]

1

In [128]:
x

[54, 10, 4, 3, 3, 2, 1, -34]

In [129]:
x[0]

54

In [130]:
x[0] = 12

In [131]:
x

[12, 10, 4, 3, 3, 2, 1, -34]

In [132]:
a[0]

1

In [134]:
#a[0] = 12

<font size="3"><br>**Exercise:** <br>Write a program that gets ’c’, ’m’ and ’p’ characters from the list below and adds it to the beginning of my_course. (Make sure that ’c’, ’m’ and ’p’ are uppercase) <br>

~~~
my_list = ["c" , "o" , "m" , "p" , "u" , "t", "e" , "r"]
my_course = "(Python for Economics)"

~~~
<br>
<font size="3"> The final output is as follows: "CMP(Python for Economics)"

In [141]:
my_list = ['c' , 'o' , 'm' , 'p' , 'u' , 't', 'e', 'r']
my_course = "(Python for Economics)"

part_1 = my_list[0] + my_list[2] + my_list[3]
part_1 = part_1.upper()
part_1

new = part_1 + my_course
print(new)

CMP(Python for Economics)


In [143]:
my_list.index('c')

0

In [144]:
my_list[0]

'c'

# 4:10

# 4. Control flow (Loop, Conditional Control)

<img src = "https://github.com/saeed-saffari/alzahra-workshop-spr2021/blob/main/lecture/PIC/Loops-in-Python.png?raw=true" width = 600 >


Control flow is the set of techniques that allows us to manage the way code is executed.
For example, you might want to execute some code provided a certain condition is `True`.
Or you might want to repeat a certain action a given number of times.
Or, again, you might want to run code as long as a condition is `True`, but you do not know how many times the code should be executed before the condition turns `True`.

The three examples are managed by the statements [`if`](https://docs.python.org/3/reference/compound_stmts.html#the-if-statement), [`for`](https://docs.python.org/3/reference/compound_stmts.html#the-for-statement) and [`while`](https://docs.python.org/3/reference/compound_stmts.html#the-while-statement).
Here we see basic applications, just to convey the idea.
We should keep in mind two things when working with control flow statements:
- The control flow statement has to be on a line that end with a semicolon (`:`).
- The code block following the control flow statements has to be indented with four whitespaces.

The `if` and `while` statements work provided that a certain condition holds.
Such condition is a boolean value that is either `True` or `False`.
Python supports boolean arithmetics with the keywords `and`, `or` and `not`.

The `if` statement can be expanded to the `if`-`elif`-`else`.


## 4.1 Loops
When working with collections (`lists`, `sets`, `dictionaries`, etc.) it is possible
to operate on every item using a loop.

Defining a loop requires an iterable (in our example it is `sample_list`), which 
could be `list`, `tuple`, `set` or any other object that supports iteration. By
specifying `item` we are asking python to iterate over one element of the list 
at a time, storing this element in a variable called `item`. 

Operations that should be perfomed within the loop must be idented (typically with
four leading spaces, but some use tabs, two spaces, or some other identation). Also
note the colon `:` at the end of the `for` statement.


### 4.1.1 for
The simplest loop is `for` loop, with the following syntax:

```python
sample_list = [1, 4, 7, 10]
for item in sample_list:
    print(item)
```

In [ ]:
for ... in ...:
    ...
    ...
    ...
    ...
    

In [145]:
users = ['Saeed', 'Sanaz', "David", 'James', "Sarah"]
users

['Saeed', 'Sanaz', 'David', 'James', 'Sarah']

In [148]:
for user in users:
    print(user)

Saeed
Sanaz
David
James
Sarah


In [153]:
for user in users:
    print("Hi")
    print(user)
print('Bye')

Hi
Saeed
Hi
Sanaz
Hi
David
Hi
James
Hi
Sarah
Bye


In [154]:
a = [1,5,8,12,48]
a

[1, 5, 8, 12, 48]

In [155]:
for i in a:
    print(i*2)

2
10
16
24
96


In [156]:
for i in [1,5,8,12,48]:
    print(i*2)

2
10
16
24
96


In [157]:
list(range(-10,10))

[-10, -9, -8, -7, -6, -5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [158]:
for i in range(3,10):
    print(i)

3
4
5
6
7
8
9


In [161]:
square = []

for i in range(1,25):
    #print(i)
    a = i ** 2
    square.append(a)

    print(square)

[1]
[1, 4]
[1, 4, 9]
[1, 4, 9, 16]
[1, 4, 9, 16, 25]
[1, 4, 9, 16, 25, 36]
[1, 4, 9, 16, 25, 36, 49]
[1, 4, 9, 16, 25, 36, 49, 64]
[1, 4, 9, 16, 25, 36, 49, 64, 81]
[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]
[1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121]
[1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144]
[1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169]
[1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196]
[1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225]
[1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256]
[1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256, 289]
[1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256, 289, 324]
[1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256, 289, 324, 361]
[1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256, 289, 324, 361, 400]
[1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256, 289, 324, 361, 400, 441]
[1, 4, 9, 16, 25, 36, 49, 64, 81, 100

### 4.1.2 while
Python `While` Loop is used to execute a block of statements repeatedly until a given condition is satisfied. And when the condition becomes false, the line immediately after the loop in the program is executed. While loop falls under the category of indefinite iteration. Indefinite iteration means that the number of times the loop is executed isn’t specified explicitly in advance. 

<img src = "https://github.com/saeed-saffari/alzahra-workshop-spr2021/blob/main/lecture/PIC/Looping.jpg?raw=true">

In [ ]:
while ...:
    ...
    ...
    ...
    ...

In [162]:
n = 0
n < 10

True

In [167]:
n = 0

while n < 10:
    print(n)
    n = n + 1   # n += 1
    #print(n)

0
1
2
3
4
5
6
7
8
9


In [171]:
1e-4

0.0001

In [170]:
x = 1
while x > 10 ** (-4):     # 1e-4:
    print(x)
    #x = x / 5
    x /= a5
    #print(x)

1
0.2
0.04
0.008
0.0016
0.00032


## 4.2 Conditional Control (if, elif, else)

<img src = "https://github.com/saeed-saffari/alzahra-workshop-spr2021/blob/main/lecture/PIC/if-else.jpg?raw=true" width = "700">

If execution of command should depend on some condition, then we can use `if/else` syntax:

```python
if some_condition:
    # run code here
elif some_other_condition:
    # run code here
else:
    # run code here
```



In [ ]:
if ...:
    ...
else:
    ...

In [173]:
3 < 2

False

In [172]:
if 3 < 2:
    print('This is True!')

In [3]:
age = int(input('Enter you age: '))

if age >= 18:
    print('You are old enought to vote!')
else:
    print("You can NOT vote yet!")
    print('You can will vote after ' + str(18 - age) + ' years!')
    print("You are now {} years old. You can vote after {} years.".format(age, 18 - age))

Enter you age: 10
You can NOT vote yet!
You can will vote after 8 years!
You are now 10 years old. You can vote after 8 years.


In [7]:
age = int(input('Enter you age: '))

if age < 5:
    price = 5
elif age < 16:
    price = 8
else:                #elif age >= 16:
    price = 20
    
print('Your cost is ${}!'.format(price))

Enter you age: 27
Your cost is $20!


In [12]:
my_list = [1,2,5,12,45,44,103,105,208]

for num in my_list:
    #print(num)
    if num % 2 == 0:
        print(num, 'is even number!')
    else:
    #elif num % 2 == 1:
    #elif num % 2 != 0:
        print(num, 'is ODD number!')

1 is ODD number!
2 is even number!
5 is ODD number!
12 is even number!
45 is ODD number!
44 is even number!
103 is ODD number!
105 is ODD number!
208 is even number!


<font size="3"><br>**Exercise:** <br>With run below command, create a list with 15 random numbers between 1 and 100. Then write a program that prints separate lists of odd and even numbers. <br>

~~~
import random
list_1 = random.sample(range(1, 100), 15)

~~~

In [18]:
import random
list_1 = random.sample(range(1, 100), 15)
list_1

[60, 68, 94, 62, 18, 29, 14, 53, 76, 67, 59, 56, 8, 96, 83]

In [19]:
odd_num = []
even_num = []

for i in list_1:
    if i % 2 == 0:
        even_num.append(i)
    else:
        odd_num.append(i)

In [21]:
print(odd_num)
print(even_num)

[29, 53, 67, 59, 83]
None


In [22]:
even_num.sort()

In [23]:
even_num

[8, 14, 18, 56, 60, 62, 68, 76, 94, 96]

# 5. Functions (`def` and `lambda`)

## 5.1 Functions (def)

<img src = "https://github.com/saeed-saffari/alzahra-workshop-spr2021/blob/main/lecture/PIC/python-functions.gif?raw=true" width = 750>

You might have noticed so far that Python is rather limited as a programming language.
This is because Python is a general-purpose language: it should be flexible enough for everybody to achieve their goals, but it cannot afford to be a BIG library of code.


We can define a new function by using the `def` keyword, listing arguments in round parentheses.
Also, if we want the function to give back something, we should explicitly instruct Python to do so with the `return` keyword.

**General structure:**
```python
def function_name(parameters):
    """ codes """
    statement(s)
```


In [ ]:
def test_1():
    ...
    ...
    ...
    return ... / print(...)

In [ ]:
test_1()

In [ ]:
def test_2(x,y,z):
    ...
    ...
    ...
    return ... / print(...)

In [ ]:
test_2(20,45,12)

For example, Python does not know what a square root of a number is.
While we can always rely on the fact that $\sqrt[n]{x} = x^{1/n}$, it might be useful to define a dedicated function that does the job.
In this example, defining a function improves code readability.

In [24]:
def sqrt(x):
    return x ** (1/2)

Here we defined the square-root function.
As it is, this function works for all values of $x$, even negative ones (the result will be a complex number).

Again, note that the `def` statement ends with a semicolon (`:`) and any code that belongs to the function is indented with respect to the line containing the keyword `def`.

In [25]:
sqrt(2)

1.4142135623730951

An example of a slightly more sophisticated function is one that computes the real roots a (non-negative) number.

This function performs the so-called _input validation_.
To ensure that the result of $\sqrt[n]{x}$ is a real number, we must ensure that $x$ is non-negative.
We do this with an `if` clause and instructing Python to terminate the execution of the function with a `raise` command, which issues an error.
Also, the function we defined takes two arguments: `x` and `n`.
By specifying `n=2` in the `def` statement, we are saying that the default value of $n$ should be two, but the user can change that value arbitrarily.

In [34]:
def real_nth_root(x, n=2):
    if x < 0:
        raise ValueError('Root argument must be non-negetive!')
    return x ** (1/n)

In [35]:
real_nth_root(2,4)

1.189207115002721

In [33]:
#real_nth_root(-2,4)

In [36]:
real_nth_root(2)

1.4142135623730951

### 5.1.1 Other example

In [37]:
def mean3(x,y,z):
    return (x+y+z) / 3

In [41]:
mean3(34,56,12)

34.0

In [42]:
def price():
    age = int(input('Enter you age: '))

    if age < 5:
        price = 5
    elif age < 16:
        price = 8
    else:                #elif age >= 16:
        price = 20
        
    print('Your cost is ${}!'.format(price))

In [43]:
price()

Enter you age: 10
Your cost is $8!


In [44]:
price()

Enter you age: 29
Your cost is $20!


**Guess number**

In [55]:
import random
cump_num = random.randint(1,100)
#cump_num

In [51]:
def guess_number():
    user_number = int(input('Enter your guess number: '))
    
    while user_number != cump_num:
        if user_number > cump_num:
            print('Enter the smaller one!')
        else:
        #elif user_number < cump_num:
            print('Enter the bigger one!')
        
        user_number = int(input('Enter your guess number: '))
        
        
        
    
    print('Well done!')

In [56]:
guess_number()

Enter your guess number: 50
Enter the smaller one!
Enter your guess number: 25
Enter the smaller one!
Enter your guess number: 12
Enter the bigger one!
Enter your guess number: 18
Enter the bigger one!
Enter your guess number: 21
Enter the smaller one!
Enter your guess number: 19
Well done!


In [57]:
# for practice, add limit!

### 5.1.2 Diffrence between return and print

In [63]:
def fun1(x1):
    print(x1 + 2)
    
def fun2(x1):
    return x1 + 2

x1 = 3


print(x1)
#print(fun1(x1))
#print(fun1(x1) * 2)


print(fun2(x1))
print(fun2(x1) * 2)

3
5
10


<font size="3"><br>**Exercise:** <br>Write a program, which asks for initial value of investment ($k_0$) and for the interest rate ($r$). Use Function ($def$) to calculate capital after n years ($k_n$) <br>

 $$ k_n = k_0\times(1 + r)^n $$

In [67]:
def capital(k_0, r, n):
    k_n = k_0 * (1+r)**n
    
    print('Your investment after {} years is equals to ${:0.3f}.'.format(n, k_n))
    
capital(200, 0.2, 10)

Your investment after 10 years is equals to $1238.347.


In [68]:
def capital(k_0, r, n):
    k_n = k_0 * (1+r)**n
    
    list_1 = []
    for i in range(0,n+1):
        k_n = k_0 * (1+r)**i
        list_1.append(round(k_n))
    
    print("Timeline: ", list_1)
    print('Your investment after {} years is equals to ${:0.3f}.'.format(n, k_n))
    
capital(200, 0.2, 10)

Timeline:  [200, 240, 288, 346, 415, 498, 597, 717, 860, 1032, 1238]
Your investment after 10 years is equals to $1238.347.


## 5.2 Functions (Lambda)

Finally, a quick-and-dirty way to define functions is by using `lambda` functions, which are also known as [anonymous functions](https://en.wikipedia.org/wiki/Anonymous_function).
The inner workings of anonymous functions are quite complicated and we will not cover them.
Here it should suffice to know that these functions are defined _in-line_, which means that we do not write a code block for them.
Anonymous functions are useful for quick one-liners that do not require much work.

We can re-define the function `sqrt` above as an anonymous function.

In [69]:
root2 = lambda x : x ** (1/2)

The syntax is as follows.
The function `root2` takes an argument `x`, which is indicated right after the keyword `lambda`.
There is a semicolon following the statment, after which we find the main task performed by the function we defined.

In [70]:
root2(2)

1.4142135623730951

### 5.2.1 Other example

In [71]:
m = lambda x,y,z: (x+y+z) / 3

In [74]:
m(12,45,-25)

10.666666666666666

<br>**Exercise:** <br>Suppose a labor works H hours in a month and gets \$16 per hour.<b>

- (a) Write a function that calculates income.<br>

- (b) Then, If we have a fixed tax of \$200, how to change income?
    
- (c) How about if our tax rate is \$1.5 per hour?

- (d) Write a program (function) that gets a H as working hours and compare the situation, and say will this labor work or not? And with which tax system will he(she) earn more?